src/Features/CSharp/Portable/CodeRefactorings/DeclareAsNullable/DeclareAsNullableCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,181 @@
+    internal class DeclareAsNullableCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var textSpan = context.Span;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var symbolToFix = await TryGetSymbolToFix(root, textSpan, document, cancellationToken).ConfigureAwait(false);
+            if (symbolToFix == null ||
+                symbolToFix.Locations.Length != 1 ||
+                !symbolToFix.IsNonImplicitAndFromSource())
+            {
+                return;
+            }
+
+            if (!IsFixableType(symbolToFix))
+            {
+                return;
+            }
+
+            var declarationLocation = symbolToFix.Locations[0];
+            var node = declarationLocation.FindNode(getInnermostNodeForTie: true, cancellationToken);
+
+            var typeToFix = TryGetTypeToFix(node);
+            if (typeToFix == null || typeToFix is NullableTypeSyntax)
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    CSharpFeaturesResources.Declare_as_nullable,
+                    c => UpdateDocumentAsync(document, typeToFix, c)));
+        }
+
+        private bool IsFixableType(ISymbol symbolToFix)
+        {
+            ITypeSymbol type = null;
+            switch (symbolToFix)
+            {
+                case IParameterSymbol parameter:
+                    type = parameter.Type;
+                    break;
+                case ILocalSymbol local:
+                    type = local.Type;
+                    break;
+                case IPropertySymbol property:
+                    type = property.Type;
+                    break;
+                case IMethodSymbol method when method.IsDefinition:
+                    type = method.ReturnType;
+                    break;
+                case IFieldSymbol field:
+                    type = field.Type;
+                    break;
+                default:
+                    return false;
+            }
+
+            return type?.IsReferenceType == true;
+        }
+
+        private TypeSyntax TryGetTypeToFix(SyntaxNode node)
+        {
+            switch (node)
+            {
+                case ParameterSyntax parameter:
+                    return parameter.Type;
+
+                case VariableDeclaratorSyntax declarator:
+                    if (declarator.IsParentKind(SyntaxKind.VariableDeclaration))
+                    {
+                        var declaration = (VariableDeclarationSyntax)declarator.Parent;
+                        return declaration.Variables.Count == 1 ? declaration.Type : null;
+                    }
+                    return null;
```

Comment:
what about things like a foreach block that doesn't have a VarDecl?

Reply:
I don't see a good scenario in a `foreach` where a null test would indicate that the iteration variable should be declared as nullable.

---
In reply to: [186297577](https://github.com/dotnet/roslyn/pull/26661#discussion_r186297577) [](ancestors = 186297577)